kore/src/Kore/Step/Simplification/SubstitutionSimplifier.hs

```diff
@@ -0,0 +1,147 @@
+{- |
+Copyright   : (c) Runtime Verification, 2018
+License     : NCSA
+
+ -}
+
+module Kore.Step.Simplification.SubstitutionSimplifier
+    ( SubstitutionSimplifier (..)
+    , simplification
+    , unification
+    ) where
+
+import Control.Error
+    ( maybeT
+    )
+import Data.Function
+    ( (&)
+    )
+import Data.Map.Strict
+    ( Map
+    )
+
+import Kore.Internal.OrPredicate
+    ( OrPredicate
+    )
+import qualified Kore.Internal.OrPredicate as OrPredicate
+import Kore.Internal.Predicate
+    ( Predicate
+    )
+import qualified Kore.Internal.Predicate as Predicate
+import Kore.Internal.TermLike
+    ( TermLike
+    )
+import qualified Kore.Predicate.Predicate as Syntax
+    ( Predicate
+    )
+import qualified Kore.Predicate.Predicate as Syntax.Predicate
+import Kore.Step.Simplification.Simplify
+    ( MonadSimplify
+    )
+import Kore.Substitute
+    ( SubstitutionVariable
+    )
+import Kore.Unification.Substitution
+    ( Substitution
+    )
+import Kore.Unification.SubstitutionNormalization
+    ( normalize
+    , normalizeSubstitution
+    )
+import qualified Kore.Unification.UnifierImpl as Unifier
+import Kore.Unification.Unify
+    ( MonadUnify
+    )
+import qualified Kore.Unification.Unify as Unifier
+import Kore.Variables.UnifiedVariable
+    ( UnifiedVariable
+    )
+
+newtype SubstitutionSimplifier simplifier =
+    SubstitutionSimplifier
+        { simplifySubstitution
+            :: forall variable
+            .  SubstitutionVariable variable
+            => Substitution variable
+            -> simplifier (OrPredicate variable)
+        }
+
+{- | A 'SubstitutionSimplifier' to use during simplification.
+
+If the 'Substitution' cannot be normalized, this simplifier moves the
+denormalized part into the predicate, but returns the normalized part as a
+substitution.
+
+ -}
+simplification :: MonadSimplify simplifier => SubstitutionSimplifier simplifier
+simplification =
+    SubstitutionSimplifier worker
+  where
+    worker substitution = maybeUnwrapSubstitution $ do
+        deduplicated <-
+            -- TODO (thomas.tuegel): If substitution de-duplication fails with a
+            -- unification error, this will still discard the entire
+            -- substitution into the predicate. Fortunately, that seems to be
+            -- rare enough to discount for now.
+            Unifier.deduplicateSubstitution substitution
+            & Unifier.maybeUnifierT
+        OrPredicate.fromPredicates <$> traverse normalize1 deduplicated
+      where
+        maybeUnwrapSubstitution =
+            let unwrapped =
+                    OrPredicate.fromPredicate
+                    . Predicate.fromPredicate
+                    . Syntax.Predicate.markSimplified
+                    -- TODO (thomas.tuegel): Promoting the entire substitution
```

Comment:
I thought that promoting cyclic stuff happens in `Predicate.fromNormalization`, not here, am I missing something? Wouldn't `unwrapped` only apply to errors from deduplicateSubstitution?

Reply:
Yes, promoting cycles happens in `fromNormalization`, this only applies if there is a unification error (`UnsupportedPatterns`) while simplifying `\and`s. I have only seen that happen in contrived tests. I think it is unlikely to happen in practice; if the patterns are unsupported for unification, how did they get into a substitution?

Reply:
Ok, thanks for the answer. That also clarifies that this TODO and the one above are talking about the same thing.

Would it make sense then to wrap only the `deduplicateSubstitution` call in `maybeUnwrapSubstitution`, and not the entire `do`?

Reply:
> Would it make sense then to wrap only the `deduplicateSubstitution` call in `maybeUnwrapSubstitution`, and not the entire `do`?

It would, but in my very next pull request, I'm actually going to remove it here entirely.